crates/shackle-ls/src/handlers/rename_symbol.rs

```diff
@@ -0,0 +1,154 @@
+use lsp_server::ResponseError;
+use lsp_types::{request::Rename, RenameParams, TextEdit, WorkspaceEdit};
+use shackle::{
+	db::CompilerDatabase,
+	file::ModelRef,
+	hir::{
+		db::Hir,
+		ids::{LocalEntityRef, NodeRef, PatternRef},
+		source::{find_node, Point},
+	},
+	syntax::db::SourceParser,
+	utils,
+};
+use std::collections::HashMap;
+
+use crate::{db::LanguageServerContext, dispatch::RequestHandler, utils::node_ref_to_location};
+
+#[derive(Debug)]
+pub struct RenameHandler;
+
+pub struct SymbolHandlerData {
+	model_ref: ModelRef,
+	cursor_pos: Point,
+	new_name: String,
+}
+
+impl RequestHandler<Rename, SymbolHandlerData> for RenameHandler {
+	fn prepare(
+		db: &mut impl LanguageServerContext,
+		params: RenameParams,
+	) -> Result<SymbolHandlerData, ResponseError> {
+		// Verify if the new name is valid
+		let mut attempted_new_name = params.new_name.chars();
+
+		// identifiers can consist of anything which is not in these chars
+		let cursor_pos = Point {
+			row: params.text_document_position.position.line as usize,
+			column: params.text_document_position.position.character as usize,
+		};
+
+		// cannot include single quotes
+		if attempted_new_name.any(|ch| ch == '\'') {
+			return Err(ResponseError {
+				code: -32602,
+				message: "Identifier cannot include single quotes".into(),
+				data: None,
+			});
+		}
+
+		// the file it is in
+		let model_ref =
+			db.set_active_file_from_document(&params.text_document_position.text_document)?;
+
+		// pretty print it to add single quotes, etc as necessary
+		let new_name = utils::pretty_print_identifier(&params.new_name);
+
+		Ok(SymbolHandlerData {
+			cursor_pos,
+			new_name,
+			model_ref,
+		})
+	}
+
+	fn execute(
+		db: &CompilerDatabase,
+		data: SymbolHandlerData,
+	) -> Result<Option<WorkspaceEdit>, ResponseError> {
+		// It's messy to write the full struct every time
+		let create_error = |msg: &str| {
+			Err(ResponseError {
+				code: -32602,
+				message: msg.into(),
+				data: None,
+			})
+		};
+
+		// Find the node that is possibly going to be changed
+		let node = match find_node(db, *data.model_ref, data.cursor_pos, data.cursor_pos) {
+			Some(n) => n,
+			None => return create_error("Identifier not selected"),
+		};
```

Comment:
Nitpick - this could be written as 
```rust
let node = find_node(db, *data.model_ref, data.cursor_pos, data.cursor_pos)
    .ok_or_else(|| create_error("Identifier not selected"))?;
```

Reply:
This in particular is something I was hoping there was a better way to do, I'll get on to implementing everything now.